src/action/impl.rs

```diff
@@ -0,0 +1,129 @@
+use std::{fmt::Debug, future::Future, pin::Pin};
+
+use serde::{Deserialize, Serialize};
+
+use super::{ActionKind, OperationTarget};
+use crate::{backend::Backend, Gateway};
+
+/// The marker trait for all action runs, this trait should not be used and is only used
+/// to make the return type of [`Gateway::run`] easily known.
+///
+/// [`Gateway::run`]: crate::Gateway::run
+pub trait ActionRunner<Success, Failure>: private::Sealed + Send {
+	#[doc(hidden)]
+	unsafe fn __run<B: Backend>(
+		self,
+		gateway: &Gateway<B>,
+	) -> Pin<Box<dyn Future<Output = Result<Success, Failure>> + Send>>;
+
+	// TODO: move `Action::validate` logic into this function.
+	#[doc(hidden)]
+	unsafe fn __validate(&self) -> Result<(), super::ActionError>;
```

Comment:
This should be public, renamed to `validate`, not marked unsafe, and each action should move validation logic to here to allow for specialization (ie if we're creating we don't need to check for data)